File: src/main/java/javabot/web/views/MainView.java
package javabot.web.views;

import com.antwerkz.sofia.Sofia;
import com.google.common.base.Charsets;
import com.google.common.base.Optional;
import com.google.inject.Injector;
import io.dropwizard.views.View;
import javabot.dao.AdminDao;
import javabot.dao.ChannelDao;
import javabot.dao.FactoidDao;
import javabot.model.Channel;
import javabot.web.JavabotConfiguration;
import javabot.web.model.InMemoryUserCache;
import javabot.web.model.User;

import javax.inject.Inject;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.WebApplicationException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public abstract class MainView extends View {

    public static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy.MM.dd hh:mm");
    private HttpServletRequest request;

    @Inject
    private AdminDao adminDao;

    @Inject
    private ChannelDao channelDao;

    @Inject
    private FactoidDao factoidDao;

    private List<String> errors = new ArrayList<>();

    public MainView() {
        super("/main.ftl", Charsets.ISO_8859_1);
    }

    public MainView(final Injector injector, final HttpServletRequest request) {
        this();
        injector.injectMembers(this);
        this.request = request;
    }

    public Sofia sofia() {
        return new Sofia();
    }

    public abstract String getChildView();

    public long getFactoidCount() {
        return factoidDao.count();
    }

    public boolean loggedIn() {
        Cookie cookie = getSessionCookie();
        return cookie != null && InMemoryUserCache.INSTANCE.getBySessionToken(cookie.getValue()).isPresent();
    }

    public boolean isAdmin() {
        Cookie cookie = getSessionCookie();
        if (cookie != null) {
            Optional<User> optional = InMemoryUserCache.INSTANCE.getBySessionToken(cookie.getValue());
            return optional.isPresent() && adminDao.getAdminByEmailAddress(optional.get().getEmail()) != null;
        } else {
            return false;
        }
    }

    public String getCurrentChannel() {
        return "";
    }

    public List<Channel> getChannels() {
        return channelDao.getChannels(isAdmin());
    }

    public String encode(String value) throws UnsupportedEncodingException {
        return java.net.URLEncoder.encode(value, "UTF-8");
    }

    private Cookie getSessionCookie() {
        for (Cookie cookie : request.getCookies()) {
            if (cookie.getName().equals(JavabotConfiguration.SESSION_TOKEN_NAME)) {
                return cookie;
            }
        }
        return null;
    }

    public HttpServletRequest getRequest() {
        return request;
    }

    public void addError(String message) {
        errors.add(message);
    }

    public List<String> getErrors() {
        return errors;
    }

    public boolean hasErrors() {
        return !errors.isEmpty();
    }
    public String format(final LocalDateTime date) {
        return DATE_TIME_FORMATTER.format(date);
    }
}

File: src/main/kotlin/javabot/web/views/MainView.kt
package javabot.web.views

import com.antwerkz.sofia.Sofia
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.ArrayList
import javabot.dao.AdminDao
import javabot.dao.ApiDao
import javabot.dao.ChannelDao
import javabot.dao.FactoidDao
import javabot.model.Channel
import javabot.model.javadoc.JavadocApi
import javabot.web.JavabotConfiguration
import javabot.web.model.InMemoryUserCache.INSTANCE
import javax.servlet.http.Cookie
import javax.servlet.http.HttpServletRequest

abstract class MainView(
    var adminDao: AdminDao,
    var channelDao: ChannelDao,
    var factoidDao: FactoidDao,
    var apiDao: ApiDao,
    val request: HttpServletRequest,
) {

    companion object {
        val DATE_TIME_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy.MM.dd hh:mm")
    }

    private val errors = ArrayList<String>()

    fun sofia(): Sofia {
        return Sofia
    }

    abstract fun getChildView(): String

    fun getFactoidCount(): Long {
        return factoidDao.count()
    }

    fun loggedIn(): Boolean {
        return INSTANCE.getBySessionToken(getSessionCookie()?.value) != null
    }

    fun isAdmin(): Boolean {
        val cookie = getSessionCookie()
        if (cookie != null) {
            val user = INSTANCE.getBySessionToken(cookie.value)
            return user != null && adminDao.getAdminByEmailAddress(user.email) != null
        } else {
            return false
        }
    }

    fun getCurrentChannel(): String {
        return ""
    }

    fun getChannels(): List<Channel> {
        return channelDao.getChannels(isAdmin())
    }

    fun getAPIs(): List<JavadocApi> {
        return apiDao.findAll()
    }

    fun encode(value: String): String {
        return java.net.URLEncoder.encode(value, "UTF-8")
    }

    private fun getSessionCookie(): Cookie? {
        return request.cookies?.firstOrNull { it.name == JavabotConfiguration.SESSION_TOKEN_NAME }
    }

    fun addError(message: String) {
        errors.add(message)
    }

    fun getErrors(): List<String> {
        return errors
    }

    fun hasErrors(): Boolean {
        return !errors.isEmpty()
    }

    open fun format(date: LocalDateTime?): String {
        return if (date != null) DATE_TIME_FORMATTER.format(date) else ""
    }

    /** Convert this view to a model map for FreeMarker */
    open fun toModel(): Map<String, Any?> {
        return mapOf(
            "sofia" to sofia(),
            "factoidCount" to getFactoidCount(),
            "loggedIn" to loggedIn(),
            "isAdmin" to isAdmin(),
            "currentChannel" to getCurrentChannel(),
            "channels" to getChannels(),
            "apis" to getAPIs(),
            "errors" to getErrors(),
            "hasErrors" to hasErrors(),
            "childView" to getChildView(),
        )
    }
}
